numpy/lib/mixins.py

```diff
@@ -0,0 +1,101 @@
+    def func(self, other):
+        try:
+            if other.__array_ufunc__ is None:
+                return NotImplemented
+        except AttributeError:
+            pass
+        return self.__array_ufunc__(ufunc, '__call__', other, self)
+    return func
+
+
+def _inplace_binary_method(ufunc):
+    def func(self, other):
+        result = self.__array_ufunc__(
+            ufunc, '__call__', self, other, out=(self,))
+        if result is NotImplemented:
+            raise TypeError('unsupported operand types for in-place '
+                            'arithmetic: %s and %s'
+                            % (type(self).__name__, type(other).__name__))
+        return result
+    return func
+
+
+def _numeric_methods(ufunc):
+    return (_binary_method(ufunc),
+            _reflected_binary_method(ufunc),
+            _inplace_binary_method(ufunc))
+
+
+def _unary_method(ufunc):
+    def func(self):
+        return self.__array_ufunc__(ufunc, '__call__', self)
+    return func
+
+
+class NDArrayOperatorsMixin(object):
+    """Implements all special methods using __array_ufunc__.
+
+    Caveats: the rarely used `divmod` (``__divmod__``) and unary `+`
+    (``__pos__``) operators do not have corresponding ufuncs. Hence this mixin
+    passes the builtin functions ``divmod`` and ``operator.pos`` to
+    ``__array_ufunc__`` instead of true ufuncs. If you inherit from this mixin,
+    your implementation needs to be able to handle this or you should override
+    these methods.
+    """
+
+    # comparisons don't have reflected and in-place versions
+    __lt__ = _binary_method(np.less)
+    __le__ = _binary_method(np.less_equal)
+    __eq__ = _binary_method(np.equal)
+    __ne__ = _binary_method(np.not_equal)
+    __gt__ = _binary_method(np.greater)
+    __ge__ = _binary_method(np.greater_equal)
+
+    # numeric methods
+    __add__, __radd__, __iadd__ = _numeric_methods(np.add)
+    __sub__, __rsub__, __isub__ = _numeric_methods(np.subtract)
+    __mul__, __rmul__, __imul__ = _numeric_methods(np.multiply)
+    __matmul__, __rmatmul__, __imatmul__ = _numeric_methods(np.matmul)
+    __div__, __rdiv__, __idiv__ = _numeric_methods(np.divide)  # Python 2 only
+    __truediv__, __rtruediv__, __itruediv__ = _numeric_methods(np.true_divide)
+    __floordiv__, __rfloordiv__, __ifloordiv__ = _numeric_methods(
+        np.floor_divide)
+    __mod__, __rmod__, __imod__ = _numeric_methods(np.mod)
+    __divmod__, __rdivmod__, __idivmod__ = _numeric_methods(divmod)
+    # TODO: handle the optional third argument for __pow__?
+    __pow__, __rpow__, __ipow__ = _numeric_methods(np.power)
+    __lshift__, __rlshift__, __ilshift__ = _numeric_methods(np.left_shift)
+    __rshift__, __rrshift__, __irshift__ = _numeric_methods(np.right_shift)
+    __and__, __rand__, __iand__ = _numeric_methods(np.logical_and)
+    __xor__, __rxor__, __ixor__ = _numeric_methods(np.logical_xor)
+    __or__, __ror__, __ior__ = _numeric_methods(np.logical_or)
```

Comment:
This is all super nice!

I do wonder whether we can do this base class in C though, so that this logic isn't duplicate inside the numpy C code.